source/Pages/jupyter_notebooks.rst

```diff
@@ -7,6 +8,147 @@ Jupyter Notebooks
      * how to use Jupyter notebooks on Spider
      * which available flavors to choose
 
-Coming soon ..
+Two methods of running jupyter notebooks in jupyter lab are discussed here: with virtual environments and with singularity containers. Some of this has also been covered in :ref:`the compute section <compute-on-spider>`.
+
+======================
+Where to run notebooks
+======================
+
+Interactive notebooks should be run on the partition machines mentioned in :ref:`prepare-workloads` and **not** on the UI machines. Building the environments and containers can be done on the UI, but once you start to run your code, please connect to a machine interactively with:
+
+.. code-block:: bash
+
+   srun --partition=short --time=12:00:00 --pty bash -i -l
+
+Which will open an interactive session on a machine in the short partition for 12 hours. In this way, the other users on the UI machines will not be disadvantaged by resources being used up by the notebook users.
+
+
+===================
+Virtual environment 
+===================
+
+Starting with *virtual environments*, these are a contained python environment you can create and load that has all the python modules and packages installed that the user needs. This ensures no componentes leak into the system environment. 
+
+You can create a virtual environment (or venv) at a path by doing:
+
+.. code-block:: bash
+   
+   python3.9 -m venv test_venv/
+
+This will create a folder called ``test_venv`` which contains the entire python environment. You can also use other python versions if you prefer. To load this environment run:
+
+.. code-block:: bash
+   
+   source test_venv/bin/activate
+
+This will show in some shells a ``(test_venv)`` next to your command line. In the environment you can now install packages using pip:
+
+.. code-block:: bash
+   
+   pip install jupyterlab pandas uproot4
```

Comment:
why pandas and uproot4?

Reply:
pandas has a nice dependency tree, so installing shows its not just 1 single package. uproot4 is arbitrary